RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,547 @@
+        int64_t s2 = s1 << 1;
+        for (int64_t j = 0; j < n; j += s2) {
+            for (int64_t k = 0; k < s1; ++k) {
+                bool b1 = (j + k) * num_rows + row < num_cols * num_rows;
+                bool b2 = (j + k + s1) * num_rows + row < num_cols * num_rows;
+                T u = b1 ? buf[(j + k) * num_rows + row] : workspace_buf[(j + k) * num_rows + row - num_rows * num_cols];
+                T v = b2 ? buf[(j + k + s1) * num_rows + row] : workspace_buf[(j + k + s1) * num_rows + row - num_rows * num_cols];
+                if(b1 && b2) {
+                    buf[(j + k) * num_rows + row] = u + v;
+                    buf[(j + k + s1) * num_rows + row] = u - v;
+                }
+                else if(!b2 && b1) {
+                    buf[(j + k) * num_rows + row] = u + v;
+                    workspace_buf[(j + k + s1) * num_rows + row - num_rows * num_cols] = u - v;
+                }
+                else if(!b2 && !b1) {
+                    workspace_buf[(j + k) * num_rows + row - num_rows * num_cols] = u + v;
+                    workspace_buf[(j + k + s1) * num_rows + row - num_rows * num_cols] = u - v;
+                }
+            }
+        }
+    }
+}
+}
+
+template <typename T>
+void fht_dispatch(
+    bool left, // Pre-multiplying?
+    blas::Layout layout,
+    int64_t num_rows,
+    int64_t num_cols,
+    int64_t log_n,
+    T* A,
+    int64_t workspace_ld=0, // leading dimension of the workspace buffer
+    T* workspace_buf=nullptr
+) {
+if(left && layout == blas::Layout::ColMajor)
+    fht_left_col_major(A, workspace_buf, workspace_ld, log_n, num_rows, num_cols);
+else if(left && layout == blas::Layout::RowMajor)
+    fht_left_row_major(A, workspace_buf, log_n, num_rows, num_cols);
+else if(!left && layout == blas::Layout::ColMajor)
+    fht_right_col_major(A, workspace_buf, log_n, num_rows, num_cols);
+else
+    fht_right_row_major(A, workspace_buf, workspace_ld, log_n, num_rows, num_cols);
+}
+}
+
+
+namespace RandBLAS::trig {
+
+/*
+ * A class for organizing all data required by the \PiRHT
+ * while applying the transform
+ */
+template <typename T>
+struct PaddedMatrix {
+    int64_t n_rows;
+    int64_t n_cols;
+    blas::Layout layout;
+    T* data; // Data-matrix
+    int64_t lda;
+    T* work; // Extra space of sz: (2 ** n_closest - lda) * cols/rows
+                  // Memory **has** to be managed by the user
+                  //NOTE: I am also going to assume that the user is kind enough to
+                  // populate this fully with zeros
+    //NOTE: We could (should?) explore the possibility of managing the buffer by ourselves
+    int64_t work_ld;
+    bool left;
+
+    // Trivial constructor
+    PaddedMatrix(
+        bool left,
+        blas::Layout layout,
+        int64_t rows,
+        int64_t cols,
+        T* data,
+        int64_t lda,
+        T* work
+    ) : left(left), layout(layout), n_rows(rows), n_cols(cols), data(data), lda(lda), work(work) {};
+};
```

Comment:
This needs more comprehensive documentation. Compare to ``SparseSkOp``. Add web docs:
<img width="300" alt="image" src="https://github.com/user-attachments/assets/cbe216f6-b497-49e4-a889-ccf7d78d5d6e" />